feat(rbac): update operator-controlled RBAC objects

ebaron opened the pull request:
This is mostly refactoring RBAC reconciling logic into a separate `rbac.go` file. This PR also ensures that Roles, RoleBindings and ClusterRoleBindings created by the operator are kept up-to-date across operator upgrades, and includes new test cases to exercise this. There are a couple of other changes here:
1. The reports pod now uses the "default" Service Account instead of the Cryostat one. The reports generator doesn't require access to the Kubernetes API, so the elevated permissions given to Cryostat aren't necessary here.
2. Instead of controlling the value of `AutomountServiceAccountToken` in the Service Account, we simply set this to true for the Cryostat pod template. Cryostat depends on its Service Account token being mounted upon startup. The definition in the Pod spec takes priority over any setting in the Service Account. [1]

Fixes: #409 

[1] https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server

andrewazores reviewed (commented):
LGTM

tthvo commented on internal/controllers/rbac.go:
```diff
@@ -0,0 +1,312 @@
+}
+
+func (r *CryostatReconciler) deleteClusterRoleBinding(ctx context.Context, cr *operatorv1beta1.Cryostat) error {
+	clusterBinding := newClusterRoleBinding(cr) // TODO move to finalizeRBAC, add tests
```
Is this still a `TODO`?

ebaron replied:
Nope, thanks for noticing that. I've removed it.

tthvo commented on internal/controllers/cryostat_controller_test.go:
```diff
@@ -2152,8 +2228,9 @@ func (t *cryostatTestInput) checkReportsDeployment() {
-	// Check that the proper Service Account is set
-	Expect(template.Spec.ServiceAccountName).To(Equal("cryostat"))
+	// Check that the default Service Account is used
+	Expect(template.Spec.ServiceAccountName).To(BeEmpty())
```
This is only specific to `FakeClient`, right? In actual cluster, it would be set to `default`?

ebaron replied:
This is for the real cluster too. When the `serviceAccountName` is left empty, the default service account is used: https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#use-the-default-service-account-to-access-the-api-server

tthvo replied:
Ahh right! I just realized golang string default-value is empty. Thanks for explaining that :D

tthvo commented on internal/controllers/cryostat_controller_test.go:
```diff
@@ -337,7 +337,83 @@ var _ = Describe("CryostatController", func() {
+				Expect(metav1.IsControlledBy(binding, cr)).To(BeTrue())
+
+				// Labels are unaffected
+				Expect(binding.Labels).To(Equal(binding.Labels))
```
I think the second one should be `oldBinding`?

ebaron replied:
Good catch! This would have tested nothing otherwise.

tthvo commented on internal/controllers/cryostat_controller_test.go:
```diff
@@ -337,7 +337,83 @@ var _ = Describe("CryostatController", func() {
+				Expect(err).ToNot(HaveOccurred())
+
+				// Labels are unaffected
+				Expect(binding.Labels).To(Equal(binding.Labels))
```
should be `oldBinding`?

tthvo commented on internal/controllers/cryostat_controller_test.go:
```diff
@@ -337,7 +337,83 @@ var _ = Describe("CryostatController", func() {
+				Expect(metav1.IsControlledBy(role, cr)).To(BeTrue())
+
+				// Labels are unaffected
+				Expect(role.Labels).To(Equal(oldRole.Labels))
```
Should we check `annotations` too?

ebaron replied:
Doesn't hurt to include those as well.

tthvo reviewed (commented):
Looks great and worked as expected! Just a couple of comments. Also, I wonder if these steps should always be applied to Kubernetes Context too?

ebaron commented:
@tthvo I addressed your feedback and added the tests to the Kubernetes branch as well. I have mostly avoided duplicating the tests if there are no logic changes between OpenShift and Kubernetes, but there actually is one in this case. The OAuth annotation is only present on OpenShift. I'm hoping after the release we can tidy up this test class a bit, it's very long. Perhaps we can avoid some copy-paste by using this `DescribeTable` Ginkgo feature: https://onsi.github.io/ginkgo/#table-specs?

tthvo commented:
> @tthvo I addressed your feedback and added the tests to the Kubernetes branch as well. I have mostly avoided duplicating the tests if there are no logic changes between OpenShift and Kubernetes, but there actually is one in this case. The OAuth annotation is only present on OpenShift. I'm hoping after the release we can tidy up this test class a bit, it's very long. Perhaps we can avoid some copy-paste by using this `DescribeTable` Ginkgo feature: https://onsi.github.io/ginkgo/#table-specs?

Right! I was wondering that too. This Ginkgo feature would help a lot! I think we could file an issue for this?

ebaron commented:
> > @tthvo I addressed your feedback and added the tests to the Kubernetes branch as well. I have mostly avoided duplicating the tests if there are no logic changes between OpenShift and Kubernetes, but there actually is one in this case. The OAuth annotation is only present on OpenShift. I'm hoping after the release we can tidy up this test class a bit, it's very long. Perhaps we can avoid some copy-paste by using this `DescribeTable` Ginkgo feature: https://onsi.github.io/ginkgo/#table-specs?
> 
> Right! I was wondering that too. This Ginkgo feature would help a lot! I think we could file an issue for this?

Sure, I filed #461. Any other concerns with this PR?

tthvo reviewed (approved):
Looks great to me :D
